packages/l10n/src/localize.ts

```diff
@@ -0,0 +1,131 @@
+ * format of `LocalizedField` to a `LocalizedString` object.
+ * The existing "localized" fields (the list version) will be removed.
+ *
+ * @param objectWithLocalizedFields
+ * the object with `LocalizedField` fields
+ * that need to be transformed into `LocalizedStrings`
+ * @param fieldNames
+ * An array of objects with following shape:
+ *   * `from`: the field to transform and to remove after
+ *   * `to`: the target field to write the transformed shape
+ */
+export const applyTransformedLocalizedFields = <
+  Input extends Record<string, unknown>,
+  Output extends Record<string, unknown>
+>(
+  objectWithLocalizedFields: Input,
+  fieldNames: FieldNameTranformationMapping[]
+): Output => {
+  const transformedFieldDefinitions = fieldNames.reduce(
+    (nextTransformed, fieldName) => ({
+      ...nextTransformed,
+      [fieldName.to]: transformLocalizedFieldToLocalizedString(
+        objectWithLocalizedFields[fieldName.from] as
+          | LocalizedField[]
+          | undefined
+      ),
+    }),
+    {}
+  );
+  const namesToOmit = fieldNames.map((fieldName) => fieldName.from);
+  const objectWithouLocalizedFields = omit<Input>(
+    objectWithLocalizedFields,
+    namesToOmit
+  );
+  return {
+    ...objectWithouLocalizedFields,
+    ...transformedFieldDefinitions,
+  } as Output;
+};
+
+/**
+ * Translates a localized string on an entity.
+ *
+ * The `localize` function receives a complete entity that can have several
+ * localized fields.
+ *
+ * Arguments
+ *  - `obj`: that entity
+ *  - `key`: the field within `obj` that might contain a localized strings
+ *  - `locale`: the language key that should be the first choice to show
+ *  - `fallbackOrder`: an array of language keys which will be tried in the
+ *     provided order for any set value
+ *  - `fallback`: the final fallback that should be displayed as a last resort.
+ *     This fallback will also be shown in case the field does not exist on the
+ *     provided object.
+ *
+ * Before `fallback` kicks in, the following is tried to display a meaningful value:
+ *  - if `locale` is `<language>-<extlang>`, eg. `de-AT`, try if `de` is set
+ *  - if not, iterate through all languages of project-settings
+ *    (passed as `fallbackOrder`) and pick the first one with a value
+ *  - if nothing is found, go through all the languages in provided localized
+ *    string an pick the first with a value
+ *  - if still no value is found display `fallback`
+ *
+ * NOTE: It is known that this might lead to strings displayed in different
+ *       languages within the same page. This is an accepted downside.
+ *
+ * NOTE: A missing field is treated like a localied string with no translations:
+ *       let a = localize({ obj: { name: { en: '', de: '' } }, locale: 'en' })
+ *       let b = localize({ obj: {}, locale: 'en' })
+ *       let c = localize({ obj: undefined, locale: 'en' })
+ *       a === b && a === c -> true
+ */
+export const localize = <Input extends Record<string, unknown>>({
```

Comment:
I wonder if we should take the chance to rename this function, as it's a bit too generic in my opinion.

```suggestion
export const formatLocalizedString = <Input extends Record<string, unknown>>({
```

The name `format*` is inspired by the react-intl format functions, which is a similar use case of what we're trying to do here.

Reply:
good call. I changed that now 👍